lib/IDE/Refactoring.cpp

```diff
@@ -4830,10 +4900,66 @@ struct CallbackClassifier {
     CurrentBlock->addPossibleCommentLoc(endCommentLoc);
   }
 
-  /// Given a callback condition, classify it as a success or failure path, or
-  /// \c None if it cannot be classified.
+  /// Whether the provided ASTNode has a child expression that force unwraps
+  /// the error parameter. Note that this doesn't walk into new scopes.
+  bool hasForceUnwrappedErrorParam(ASTNode Node) {
+    if (IsResultParam || !ErrParam)
+      return false;
+
+    class ErrUnwrapFinder : public ASTWalker {
+      const ParamDecl *ErrParam;
+      bool FoundUnwrap = false;
+
+    public:
+      explicit ErrUnwrapFinder(const ParamDecl *ErrParam)
+          : ErrParam(ErrParam) {}
+      bool foundUnwrap() const { return FoundUnwrap; }
+
+      std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+        // Don't walk into ternary conditionals as they may have additional
+        // conditions such as err != nil that make a force unwrap now valid.
+        if (isa<IfExpr>(E))
+          return {false, E};
+
+        auto *FVE = dyn_cast<ForceValueExpr>(E);
+        if (!FVE)
+          return {true, E};
+
+        auto *DRE = dyn_cast<DeclRefExpr>(FVE->getSubExpr());
+        if (!DRE)
+          return {true, E};
+
+        if (DRE->getDecl() != ErrParam)
+          return {true, E};
+
+        // If we find the node we're looking for, make a note of it, and abort
+        // the walk.
+        FoundUnwrap = true;
+        return {false, nullptr};
+      }
+
+      std::pair<bool, Stmt *> walkToStmtPre(Stmt *S) override {
+        // Don't walk into new explicit scopes, we only want to consider force
+        // unwraps in the immedidate conditional body.
```

Comment:
Typo (should be "immediate")

Reply:
Also, does this need to be limited to the immediate body?

Reply:
Also replied to this below, but the idea of restricting it to the immediate body is so we don't accidentally consider bits of code that take place in nested scopes that may have additional information about whether they can force unwrap the error.